// throat/index.js
var promise = require('promises-a');

module.exports = throttler;
function throttler(concurrency) {
  var inProgress = 0;
  var queue = [];
  function throttle(fn) {
    if (typeof fn != 'function') {
      return reject(new TypeError(fn + ' is not a function.'));
    }
    if (idle() > 0) {
      var result = fn();
      inProgress++;
      result.then(null, function () {})
            .then(function () {
              inProgress--;
              if (queue.length) {
                queue.shift()();
              }
            });
      return result;
    } else {
      var def = promise();
      queue.push(function () {
        def.fulfill(throttle(fn));
      });
      return def.promise;
    }
  }
  throttle.workers = workers;
  function workers(val) {
    if (typeof val === 'number')
      return concurrency = val;
    else return concurrency;
  }
  throttle.running = running;
  function running() {
    return inProgress;
  }
  throttle.idle = idle;
  function idle() {
    return concurrency - inProgress;
  }

  throttle.workOn = function (source, worker) {
    var def = promise();
    function nextJob() {
      throttle(function () {
        return source()
          .then(function (job) {
            nextJob();
            return worker(job);
          });
      })
      .then(null, def.reject);
    }
    nextJob();
    return def.promise;
  };
  return throttle;
}

function reject(reason) {
  var p = promise();
  p.reject(reason);
  return p.promise;
}
